File: Android/app/src/main/java/com/auralis/crisisconnect/messaging/ChannelAttachments.kt
package com.auralis.crisisconnect.messaging

import android.content.Context
import android.util.Base64
import com.auralis.crisisconnect.security.Crypto
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageMetadata
import java.io.File
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

/** A ready-to-send attachment: an already-prepared blob (compressed image / recorded voice / raw file). */
data class PendingChannelAttachment(
    val bytes: ByteArray,
    val name: String,
    val mime: String,
    val width: Int? = null,
    val height: Int? = null,
    /** Playback length in seconds, for voice notes (audio types). */
    val durationSec: Int? = null,
)

/** A decoded attachment descriptor read off a message doc; its blob is still encrypted in Storage. */
data class ChannelAttachment(
    val path: String,
    val nonce: String,
    /** Per-file AES key. Null only for legacy blobs sealed with the channel key. */
    val key: String? = null,
    val name: String,
    val mime: String,
    val size: Long,
    val width: Int? = null,
    val height: Int? = null,
    val durationSec: Int? = null,
) {
    val isImage: Boolean get() = mime.startsWith("image/")
    val isAudio: Boolean get() = mime.startsWith("audio/")

    companion object {
        /** Serializes descriptors to a plain (unencrypted) JSON array for the local Room cache. */
        fun toJsonArray(list: List<ChannelAttachment>): String {
            if (list.isEmpty()) return "[]"
            val arr = JSONArray()
            for (a in list) {
                val d = JSONObject()
                    .put("path", a.path)
                    .put("nonce", a.nonce)
                    .put("name", a.name)
                    .put("mime", a.mime)
                    .put("size", a.size)
                a.key?.let { d.put("key", it) }
                a.width?.let { d.put("width", it) }
                a.height?.let { d.put("height", it) }
                a.durationSec?.let { d.put("duration", it) }
                arr.put(d)
            }
            return arr.toString()
        }

        /** Inverse of [toJsonArray]; tolerant of malformed/empty input. */
        fun fromJsonArray(json: String?): List<ChannelAttachment> {
            if (json.isNullOrBlank()) return emptyList()
            return runCatching {
                val arr = JSONArray(json)
                buildList {
                    for (i in 0 until arr.length()) {
                        val d = arr.optJSONObject(i) ?: continue
                        val path = d.optString("path")
                        val nonce = d.optString("nonce")
                        if (path.isBlank() || nonce.isBlank()) continue
                        add(
                            ChannelAttachment(
                                path = path,
                                nonce = nonce,
                                key = d.optString("key").takeIf(String::isNotBlank),
                                name = d.optString("name", "file"),
                                mime = d.optString("mime", "application/octet-stream"),
                                size = d.optLong("size"),
                                width = if (d.has("width")) d.optInt("width") else null,
                                height = if (d.has("height")) d.optInt("height") else null,
                                durationSec = if (d.has("duration")) d.optInt("duration") else null,
                            ),
                        )
                    }
                }
            }.getOrDefault(emptyList())
        }
    }
}

/**
 * Byte-compatible port of the web dashboard's `lib/messaging/attachments.ts`. Each new blob is
 * AES-256-GCM sealed with its own random key (AAD = its Storage path) and uploaded as ciphertext at
 * `messageAttachments/{uid}/{uuid}`; a small descriptor array
 * `{path,nonce,name,mime,size,width?,height?,duration?}` is itself encrypted and stored inside the
 * message doc as `attMeta`/`attMetaNonce`, so filenames and storage paths never leak. A channel member
 * decrypts the descriptor, downloads the blob and decrypts it locally — matching what the web renders.
 */
object ChannelAttachments {
    const val MAX_ATTACHMENT_BYTES = 25L * 1024 * 1024
    const val AUTHORITY_MLS_BLOB_MIME = "application/x-crisisconnect-authority-mls-attachment"
    const val AUTHORITY_MLS_ENVELOPE_MIME = "application/x-crisisconnect-authority-mls-envelope"
    private val CACHE_MAGIC = byteArrayOf(0x43, 0x43, 0x41, 0x54, 0x54, 0x02)
    private val AUTHORITY_CONVERSATION_ID = Regex("^am2_[A-Za-z0-9_-]{43}$")
    private val AUTHORITY_ATTACHMENT_PATH = Regex(
        "^authorityMessageAttachments/am2_[A-Za-z0-9_-]{43}/[^/]{1,256}/" +
            "[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-4[0-9A-Fa-f]{3}-[89AaBb][0-9A-Fa-f]{3}-[0-9A-Fa-f]{12}$"
    )

    private fun aad(aadString: String) = aadString.toByteArray(StandardCharsets.UTF_8)

    private fun encryptBytes(key: ByteArray, aadString: String, bytes: ByteArray): Pair<String, ByteArray> {
        val nonce = Crypto.randomBytes(12)
        val cipher = Crypto.aesGcmEncrypt(key = key, nonce = nonce, plaintext = bytes, associatedData = aad(aadString))
        return Base64.encodeToString(nonce, Base64.NO_WRAP) to cipher
    }

    private fun decryptBytes(key: ByteArray, aadString: String, nonceBase64: String, cipher: ByteArray): ByteArray =
        Crypto.aesGcmDecrypt(
            key = key,
            nonce = Base64.decode(nonceBase64, Base64.NO_WRAP),
            ciphertextAndTag = cipher,
            associatedData = aad(aadString),
        )

    /**
     * Prepares MLS-v2 file descriptors. Each blob has an independent AES-256-GCM key carried only
     * inside MLS plaintext, while the Storage path binds access to the exact conversation.
     */
    suspend fun prepareAuthorityMlsAttachments(
        context: Context,
        conversationId: String,
        pendings: List<PendingChannelAttachment>,
    ): List<ChannelAttachment> = withContext(Dispatchers.IO) {
        require(AUTHORITY_CONVERSATION_ID.matches(conversationId)) { "Invalid Authority MLS conversation." }
        require(pendings.size <= 8) { "Too many Authority MLS attachments." }
        if (pendings.isEmpty()) return@withContext emptyList()
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?.takeIf { it.length in 1..256 && '/' !in it }
            ?: throw IllegalStateException("Not signed in")
        pendings.map { pending ->
                require(pending.bytes.isNotEmpty() && pending.bytes.size.toLong() <= MAX_ATTACHMENT_BYTES) {
                    "Authority MLS attachment size is invalid."
                }
                require(pending.name.toByteArray(StandardCharsets.UTF_8).size in 1..255 &&
                    pending.mime.toByteArray(StandardCharsets.UTF_8).size in 1..255
                ) { "Authority MLS attachment metadata is invalid." }
                val path = "authorityMessageAttachments/$conversationId/$uid/${UUID.randomUUID()}"
                val fileKey = Crypto.randomBytes(32)
                val (nonce, cipher) = encryptBytes(fileKey, path, pending.bytes)
                check(cacheAuthorityMlsCiphertext(context, path, cipher)) {
                    "Authority MLS attachment ciphertext could not be cached."
                }
                ChannelAttachment(
                    path = path,
                    nonce = nonce,
                    key = Base64.encodeToString(fileKey, Base64.NO_WRAP),
                    name = pending.name,
                    mime = pending.mime,
                    size = pending.bytes.size.toLong(),
                    width = pending.width,
                    height = pending.height,
                    durationSec = pending.durationSec,
                )
            }
    }

    /** Uploads the exact locally cached ciphertext; safe to retry after an offline BLE send. */
    suspend fun ensureAuthorityMlsAttachmentsUploaded(
        context: Context,
        attachments: List<ChannelAttachment>,
    ) = withContext(Dispatchers.IO) {
        if (attachments.isEmpty()) return@withContext
        val uid = FirebaseAuth.getInstance().currentUser?.uid
            ?.takeIf { it.length in 1..256 && '/' !in it }
            ?: throw IllegalStateException("Not signed in")
        attachments.forEach { attachment ->
            require(AUTHORITY_ATTACHMENT_PATH.matches(attachment.path) &&
                attachment.path.split('/').getOrNull(2) == uid && !attachment.key.isNullOrBlank()) {
                "Authority MLS attachment upload binding is invalid."
            }
            val cipher = readCachedAuthorityMlsCiphertext(context, attachment.path)
                ?: throw IllegalStateException("Authority MLS attachment ciphertext is unavailable.")
            val reference = FirebaseStorage.getInstance().reference.child(attachment.path)
            val digest = Base64.encodeToString(
                MessageDigest.getInstance("SHA-256").digest(cipher),
                Base64.NO_WRAP,
            )
            suspend fun uploadedObjectMatches(): Boolean {
                val metadata = runCatching { reference.metadata.await() }.getOrNull() ?: return false
                if (metadata.sizeBytes != cipher.size.toLong()) return false
                val storedDigest = metadata.getCustomMetadata("cc-sha256")
                if (storedDigest != null) return storedDigest == digest
                return runCatching {
                    reference.getBytes(MAX_ATTACHMENT_BYTES + 4096).await().contentEquals(cipher)
                }.getOrDefault(false)
            }
            val metadata = StorageMetadata.Builder()
                .setContentType("application/octet-stream")
                .setCustomMetadata("cc-sha256", digest)
                .build()
            runCatching { reference.putBytes(cipher, metadata).await() }
                .getOrElse { uploadError ->
                    // Storage objects are immutable. A prior successful upload followed by an app
                    // crash is therefore success only when the exact ciphertext already exists.
                    if (!uploadedObjectMatches()) throw uploadError
                }
        }
    }

    /** Stores only opaque AES-GCM ciphertext in the app-private cache. */
    fun cacheAuthorityMlsCiphertext(context: Context, path: String, cipher: ByteArray): Boolean {
        if (!AUTHORITY_ATTACHMENT_PATH.matches(path) || cipher.size !in 17..(MAX_ATTACHMENT_BYTES.toInt() + 16)) {
            return false
        }
        val cacheFile = mediaCacheFile(context, path)
        return runCatching {
            val tmp = File.createTempFile("media", ".tmp", cacheFile.parentFile)
            tmp.writeBytes(CACHE_MAGIC + cipher)
            if (!tmp.renameTo(cacheFile)) {
                tmp.copyTo(cacheFile, overwrite = true)
                tmp.delete()
            }
            true
        }.getOrDefault(false)
    }

    fun readCachedAuthorityMlsCiphertext(context: Context, path: String): ByteArray? {
        if (!AUTHORITY_ATTACHMENT_PATH.matches(path)) return null
        val encoded = runCatching { mediaCacheFile(context, path).readBytes() }.getOrNull() ?: return null
        if (encoded.size <= CACHE_MAGIC.size ||
            !encoded.copyOfRange(0, CACHE_MAGIC.size).contentEquals(CACHE_MAGIC)) return null
        return encoded.copyOfRange(CACHE_MAGIC.size, encoded.size)
            .takeIf { it.size in 17..(MAX_ATTACHMENT_BYTES.toInt() + 16) }
    }

    /** Best-effort rollback for blobs uploaded before MLS staging fails. */
    suspend fun deleteAuthorityMlsAttachments(attachments: List<ChannelAttachment>) = withContext(Dispatchers.IO) {
        attachments
            .asSequence()
            .filter { AUTHORITY_ATTACHMENT_PATH.matches(it.path) }
            .forEach { attachment ->
                runCatching { FirebaseStorage.getInstance().reference.child(attachment.path).delete().await() }
            }
    }

    /**
     * Encrypts + uploads every pending blob, then returns the encrypted descriptor fields to merge into
     * the message doc (`attMeta` + `attMetaNonce`), or null when there are no attachments.
     */
    suspend fun buildAttachmentFields(
        @Suppress("UNUSED_PARAMETER") key: ByteArray,
        @Suppress("UNUSED_PARAMETER") aadString: String,
        @Suppress("UNUSED_PARAMETER") pendings: List<PendingChannelAttachment>,
    ): Map<String, Any>? {
        throw SecurityException("Legacy shared-key attachment uploads are permanently disabled.")
    }

    /** Reads + decrypts the descriptor array off a raw message doc. Empty when there are none/undecodable. */
    fun decodeAttachments(
        key: ByteArray,
        aadString: String,
        attMeta: String?,
        attMetaNonce: String?,
    ): List<ChannelAttachment> {
        if (attMeta.isNullOrBlank() || attMetaNonce.isNullOrBlank()) return emptyList()
        return runCatching {
            val bytes = decryptBytes(key, aadString, attMetaNonce, Base64.decode(attMeta, Base64.NO_WRAP))
            val arr = JSONArray(String(bytes, StandardCharsets.UTF_8))
            buildList {
                for (i in 0 until arr.length()) {
                    val d = arr.optJSONObject(i) ?: continue
                    val path = d.optString("path")
                    val nonce = d.optString("nonce")
                    if (path.isBlank() || nonce.isBlank()) continue
                    add(
                        ChannelAttachment(
                                path = path,
                                nonce = nonce,
                                key = d.optString("key").takeIf(String::isNotBlank),
                            name = d.optString("name", "file"),
                            mime = d.optString("mime", "application/octet-stream"),
                            size = d.optLong("size"),
                            width = if (d.has("width")) d.optInt("width") else null,
                            height = if (d.has("height")) d.optInt("height") else null,
                            durationSec = if (d.has("duration")) d.optInt("duration") else null,
                        ),
                    )
                }
            }
        }.getOrDefault(emptyList())
    }

    /**
     * Returns the decrypted blob for a descriptor. The offline cache stores only the original
     * ciphertext. Pre-v2 cache entries contained plaintext; they are deleted without being returned.
     */
    suspend fun fetchAttachmentBytes(
        context: Context,
        key: ByteArray?,
        aadString: String?,
        att: ChannelAttachment,
    ): ByteArray? = withContext(Dispatchers.IO) {
        if (!AUTHORITY_ATTACHMENT_PATH.matches(att.path) || att.key.isNullOrBlank()) return@withContext null
        val cacheFile = mediaCacheFile(context, att.path)
        cacheFile.takeIf { it.isFile && it.length() > 0 }?.let { cached ->
            val encoded = runCatching { cached.readBytes() }.getOrNull()
            if (encoded != null && encoded.size > CACHE_MAGIC.size &&
                encoded.copyOfRange(0, CACHE_MAGIC.size).contentEquals(CACHE_MAGIC)
            ) {
                decryptAttachment(key, aadString, att, encoded.copyOfRange(CACHE_MAGIC.size, encoded.size))
                    ?.let { return@withContext it }
            } else {
                // Legacy plaintext cache: never surface it and remove the insecure at-rest copy.
                runCatching { cached.delete() }
            }
        }
        val cipher = FirebaseStorage.getInstance().reference
            .child(att.path)
            .getBytes(MAX_ATTACHMENT_BYTES + 4096)
            .await()
        val plain = decryptAttachment(key, aadString, att, cipher) ?: return@withContext null
        if (plain.size.toLong() != att.size) return@withContext null
        // Cache ciphertext atomically (temp + rename), never plaintext.
        cacheAuthorityMlsCiphertext(context, att.path, cipher)
        plain
    }

    private fun decryptAttachment(
        legacyChannelKey: ByteArray?,
        legacyAad: String?,
        att: ChannelAttachment,
        cipher: ByteArray,
    ): ByteArray? = runCatching {
        if (!att.key.isNullOrBlank()) {
            val fileKey = Base64.decode(att.key, Base64.NO_WRAP)
            require(fileKey.size == 32) { "Attachment key has an invalid size." }
            decryptBytes(fileKey, att.path, att.nonce, cipher)
        } else {
            val channelKey = legacyChannelKey ?: return null
            val aad = legacyAad ?: return null
            decryptBytes(channelKey, aad, att.nonce, cipher)
        }
    }.getOrNull()

    /** app-private cache path for a blob, named by SHA-256 of its Storage path (which has slashes). */
    private fun mediaCacheFile(context: Context, path: String): File {
        val dir = File(context.filesDir, "authority_media").apply { if (!exists()) mkdirs() }
        val hash = MessageDigest.getInstance("SHA-256")
            .digest(path.toByteArray(StandardCharsets.UTF_8))
            .joinToString("") { "%02x".format(it) }
        return File(dir, hash)
    }
}

File: iOS/Crisis Connect/Features/Messaging/ChannelAttachments.swift
//
//  ChannelAttachments.swift
//  Crisis Connect
//
//  Byte-compatible port of the web dashboard's `lib/messaging/attachments.ts` and Android's
//  ChannelAttachments. Each new blob is AES-256-GCM sealed with its own random key (AAD = its
//  Storage path) and uploaded
//  to Firebase Storage as opaque ciphertext at
//  `authorityMessageAttachments/{conversationId}/{uid}/{uuid}`; a small
//  descriptor array `{path,nonce,name,mime,size,width?,height?,duration?}` is itself encrypted
//  and stored inside the message doc as `attMeta`/`attMetaNonce`, so filenames and storage paths
//  never leak. A channel member decrypts the descriptor, downloads the blob and decrypts it
//  locally — matching what the web renders.
//

import CryptoKit
import FirebaseAuth
import FirebaseStorage
import Foundation

/// A ready-to-send attachment: an already-prepared blob (compressed image / recorded voice / file).
struct PendingChannelAttachment: Sendable {
    let data: Data
    let name: String
    let mime: String
    var width: Int? = nil
    var height: Int? = nil
    /// Playback length in seconds, for voice notes (audio types).
    var durationSec: Int? = nil
}

/// A decoded attachment descriptor read off a message doc; its blob is still encrypted in Storage.
struct ChannelAttachment: Equatable, Identifiable, Sendable {
    var id: String { path }
    let path: String
    let nonce: String
    /// Per-file AES key. Nil only for legacy blobs sealed with the channel key.
    var keyBase64: String? = nil
    let name: String
    let mime: String
    let size: Int
    var width: Int? = nil
    var height: Int? = nil
    var durationSec: Int? = nil

    var isImage: Bool { mime.hasPrefix("image/") }
    var isAudio: Bool { mime.hasPrefix("audio/") }
}

enum ChannelAttachmentsError: Error {
    case notSignedIn
    case cryptoFailed
}

enum ChannelAttachments {
    static let maxAttachmentBytes = 25 * 1024 * 1024
    /// Android's authenticated GATT file lane currently caps one transfer at 512 KiB. Keep the
    /// Authority MLS relay at the cross-platform minimum; larger encrypted blobs wait for cloud.
    static let authorityMlsBluetoothMaxBytes = 512 * 1024
    static let authorityMlsBlobMime = "application/x-crisisconnect-authority-mls-attachment"
    static let authorityMlsEnvelopeMime = "application/x-crisisconnect-authority-mls-envelope"
    private static let cacheMagic = Data([0x43, 0x43, 0x41, 0x54, 0x54, 0x02])
    private static let authorityConversationPattern = "^am2_[A-Za-z0-9_-]{43}$"
    private static let authorityAttachmentPathPattern = "^authorityMessageAttachments/am2_[A-Za-z0-9_-]{43}/[^/]{1,256}/[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-4[0-9A-Fa-f]{3}-[89AaBb][0-9A-Fa-f]{3}-[0-9A-Fa-f]{12}$"

    // MARK: - Crypto (AES-256-GCM, ct || tag — identical to web/Android)

    private static func encryptBytes(
        key: SymmetricKey, aad: String, bytes: Data
    ) throws -> (nonceBase64: String, cipher: Data) {
        var nonceData = Data(count: 12)
        let randomStatus = nonceData.withUnsafeMutableBytes {
            SecRandomCopyBytes(kSecRandomDefault, 12, $0.baseAddress!)
        }
        guard randomStatus == errSecSuccess else { throw ChannelAttachmentsError.cryptoFailed }
        let box = try AES.GCM.seal(
            bytes,
            using: key,
            nonce: AES.GCM.Nonce(data: nonceData),
            authenticating: Data(aad.utf8)
        )
        return (nonceData.base64EncodedString(), box.ciphertext + box.tag)
    }

    private static func decryptBytes(
        key: SymmetricKey, aad: String, nonceBase64: String, cipher: Data
    ) throws -> Data {
        guard let nonceData = Data(base64Encoded: nonceBase64), cipher.count > 16 else {
            throw ChannelAttachmentsError.cryptoFailed
        }
        let box = try AES.GCM.SealedBox(
            nonce: AES.GCM.Nonce(data: nonceData),
            ciphertext: cipher.prefix(cipher.count - 16),
            tag: cipher.suffix(16)
        )
        return try AES.GCM.open(box, using: key, authenticating: Data(aad.utf8))
    }

    // MARK: - Send side

    /// Encrypts MLS-v2 file blobs with independent random keys and returns descriptors that must be
    /// carried only inside MLS plaintext. The path lets Storage rules enforce the exact conversation.
    static func prepareAuthorityMlsAttachments(
        conversationId: String,
        pendings: [PendingChannelAttachment]
    ) async throws -> [ChannelAttachment] {
        guard conversationId.range(of: authorityConversationPattern, options: .regularExpression) != nil,
              pendings.count <= 8 else { throw ChannelAttachmentsError.cryptoFailed }
        guard !pendings.isEmpty else { return [] }
        guard let uid = Auth.auth().currentUser?.uid,
              (1...256).contains(uid.count), !uid.contains("/") else {
            throw ChannelAttachmentsError.notSignedIn
        }
        var descriptors: [ChannelAttachment] = []
        for pending in pendings {
            guard !pending.data.isEmpty, pending.data.count <= maxAttachmentBytes,
                  (1...255).contains(pending.name.utf8.count),
                  (1...255).contains(pending.mime.utf8.count) else {
                throw ChannelAttachmentsError.cryptoFailed
            }
            let path = "authorityMessageAttachments/\(conversationId)/\(uid)/\(UUID().uuidString)"
            let fileKey = SymmetricKey(size: .bits256)
            let (nonce, cipher) = try encryptBytes(key: fileKey, aad: path, bytes: pending.data)
            guard cacheAuthorityMlsCiphertext(path: path, cipher: cipher) else {
                throw ChannelAttachmentsError.cryptoFailed
            }
            descriptors.append(ChannelAttachment(
                path: path,
                nonce: nonce,
                keyBase64: fileKey.withUnsafeBytes { Data($0).base64EncodedString() },
                name: pending.name,
                mime: pending.mime,
                size: pending.data.count,
                width: pending.width,
                height: pending.height,
                durationSec: pending.durationSec
            ))
        }
        return descriptors
    }

    /// Uploads the exact locally cached ciphertext; safe to retry after an offline BLE send.
    static func ensureAuthorityMlsAttachmentsUploaded(_ attachments: [ChannelAttachment]) async throws {
        guard !attachments.isEmpty else { return }
        guard let uid = Auth.auth().currentUser?.uid,
              (1...256).contains(uid.count), !uid.contains("/") else {
            throw ChannelAttachmentsError.notSignedIn
        }
        for attachment in attachments {
            let components = attachment.path.split(separator: "/", omittingEmptySubsequences: false)
            guard attachment.path.range(of: authorityAttachmentPathPattern, options: .regularExpression) != nil,
                  components.count == 4, components[2] == Substring(uid),
                  attachment.keyBase64 != nil,
                  let cipher = cachedAuthorityMlsCiphertext(path: attachment.path) else {
                throw ChannelAttachmentsError.cryptoFailed
            }
            let reference = Storage.storage().reference(withPath: attachment.path)
            let digest = Data(SHA256.hash(data: cipher)).base64EncodedString()
            func uploadedObjectMatches() async -> Bool {
                guard let existing = try? await reference.getMetadata(),
                      existing.size == Int64(cipher.count) else { return false }
                if let storedDigest = existing.customMetadata?["cc-sha256"] {
                    return storedDigest == digest
                }
                guard let existingCipher = try? await reference.data(
                    maxSize: Int64(maxAttachmentBytes + 4096)
                ) else { return false }
                return existingCipher == cipher
            }
            let metadata = StorageMetadata()
            metadata.contentType = "application/octet-stream"
            metadata.customMetadata = ["cc-sha256": digest]
            do {
                _ = try await reference.putDataAsync(cipher, metadata: metadata)
            } catch {
                // Objects are immutable. Treat a raced/crash-recovery retry as success only after
                // verifying that the already-created object is this exact ciphertext.
                guard await uploadedObjectMatches() else { throw error }
            }
        }
    }

    /// Stores only opaque AES-GCM ciphertext in the app-private cache.
    @discardableResult
    static func cacheAuthorityMlsCiphertext(path: String, cipher: Data) -> Bool {
        guard path.range(of: authorityAttachmentPathPattern, options: .regularExpression) != nil,
              (17...(maxAttachmentBytes + 16)).contains(cipher.count) else { return false }
        do {
            try (cacheMagic + cipher).write(to: mediaCacheURL(for: path), options: .atomic)
            return true
        } catch {
            return false
        }
    }

    static func cachedAuthorityMlsCiphertext(path: String) -> Data? {
        guard path.range(of: authorityAttachmentPathPattern, options: .regularExpression) != nil,
              let encoded = try? Data(contentsOf: mediaCacheURL(for: path)),
              encoded.starts(with: cacheMagic) else { return nil }
        let cipher = Data(encoded.dropFirst(cacheMagic.count))
        return (17...(maxAttachmentBytes + 16)).contains(cipher.count) ? cipher : nil
    }

    /// Best-effort rollback for blobs uploaded before MLS staging fails. Storage rules only allow
    /// the original uploader to delete these immutable objects.
    static func deleteAuthorityMlsAttachments(_ attachments: [ChannelAttachment]) async {
        for attachment in attachments where
            attachment.path.range(of: authorityAttachmentPathPattern, options: .regularExpression) != nil {
            try? await Storage.storage().reference(withPath: attachment.path).delete()
        }
    }

    /// Encrypts + uploads every pending blob, then returns the encrypted descriptor fields to merge
    /// into the message doc (`attMeta` + `attMetaNonce`), or nil when there are no attachments.
    static func buildAttachmentFields(
        key _: SymmetricKey,
        aad _: String,
        pendings _: [PendingChannelAttachment]
    ) async throws -> [String: Any]? {
        throw ChannelAttachmentsError.cryptoFailed
    }

    // MARK: - Receive side

    /// Reads + decrypts the descriptor array off a raw message doc. Empty when none/undecodable.
    static func decodeAttachments(
        key: SymmetricKey,
        aad: String,
        attMeta: String?,
        attMetaNonce: String?
    ) -> [ChannelAttachment] {
        guard let attMeta, let attMetaNonce,
              !attMeta.isEmpty, !attMetaNonce.isEmpty,
              let metaCipher = Data(base64Encoded: attMeta),
              let plain = try? decryptBytes(key: key, aad: aad, nonceBase64: attMetaNonce, cipher: metaCipher),
              let raw = try? JSONSerialization.jsonObject(with: plain) as? [[String: Any]] else {
            return []
        }
        return raw.compactMap { descriptor in
            guard let path = descriptor["path"] as? String, !path.isEmpty,
                  let nonce = descriptor["nonce"] as? String, !nonce.isEmpty else {
                return nil
            }
            return ChannelAttachment(
                path: path,
                nonce: nonce,
                keyBase64: descriptor["key"] as? String,
                name: descriptor["name"] as? String ?? "file",
                mime: descriptor["mime"] as? String ?? "application/octet-stream",
                size: (descriptor["size"] as? NSNumber)?.intValue ?? 0,
                width: (descriptor["width"] as? NSNumber)?.intValue,
                height: (descriptor["height"] as? NSNumber)?.intValue,
                durationSec: (descriptor["duration"] as? NSNumber)?.intValue
            )
        }
    }

    /// Returns the decrypted blob for a descriptor. Offline cache files contain only the original
    /// ciphertext. Pre-v2 cache entries held plaintext and are deleted without being returned.
    static func fetchAttachmentBytes(
        key: SymmetricKey?,
        aad: String?,
        attachment: ChannelAttachment
    ) async -> Data? {
        guard attachment.path.range(of: authorityAttachmentPathPattern, options: .regularExpression) != nil,
              attachment.keyBase64 != nil else { return nil }
        let cacheURL = mediaCacheURL(for: attachment.path)
        if let cached = try? Data(contentsOf: cacheURL), !cached.isEmpty {
            if cached.starts(with: cacheMagic) {
                let cipher = cached.dropFirst(cacheMagic.count)
                if let opened = decryptAttachment(
                    legacyChannelKey: key,
                    legacyAad: aad,
                    attachment: attachment,
                    cipher: Data(cipher)
                ) {
                    return opened
                }
            } else {
                // Legacy plaintext cache: remove the insecure at-rest copy and fetch ciphertext.
                try? FileManager.default.removeItem(at: cacheURL)
            }
        }
        guard let cipher = try? await Storage.storage()
            .reference(withPath: attachment.path)
            .data(maxSize: Int64(maxAttachmentBytes + 4096)) else {
            return nil
        }
        guard let plain = decryptAttachment(
            legacyChannelKey: key,
            legacyAad: aad,
            attachment: attachment,
            cipher: cipher
        ) else {
            return nil
        }
        guard plain.count == attachment.size else { return nil }
        // Cache ciphertext atomically, never plaintext.
        _ = cacheAuthorityMlsCiphertext(path: attachment.path, cipher: cipher)
        return plain
    }

    private static func decryptAttachment(
        legacyChannelKey: SymmetricKey?,
        legacyAad: String?,
        attachment: ChannelAttachment,
        cipher: Data
    ) -> Data? {
        if let encoded = attachment.keyBase64,
           let rawKey = Data(base64Encoded: encoded),
           rawKey.count == 32 {
            return try? decryptBytes(
                key: SymmetricKey(data: rawKey),
                aad: attachment.path,
                nonceBase64: attachment.nonce,
                cipher: cipher
            )
        }
        guard let legacyChannelKey, let legacyAad else { return nil }
        return try? decryptBytes(
            key: legacyChannelKey,
            aad: legacyAad,
            nonceBase64: attachment.nonce,
            cipher: cipher
        )
    }

    /// App-private cache path for a blob, named by SHA-256 of its Storage path (which has slashes).
    private static func mediaCacheURL(for path: String) -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory(), isDirectory: true)
        let directory = base.appendingPathComponent("authority_media", isDirectory: true)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let hash = SHA256.hash(data: Data(path.utf8))
            .map { String(format: "%02x", $0) }
            .joined()
        return directory.appendingPathComponent(hash, isDirectory: false)
    }
}
